sealtk/core/TimeStamp.cpp

```diff
@@ -60,69 +62,110 @@ TimeStamp& TimeStamp::set_frame_domain(int frame_domain)
+  comparator<ComparisonResult, kv::time_usec_t> compareTimes,
+  comparator<bool, kv::frame_id_t> compareFrames, bool fallback)
+{
+  bool const timeValid =
+    lhs.has_valid_time() && rhs.has_valid_time() &&
+    lhs.timeDomain == rhs.timeDomain;
+  bool const frameValid =
+    lhs.has_valid_frame() && rhs.has_valid_frame() &&
+    lhs.frameDomain == rhs.frameDomain;
+
+  if (!timeValid && !frameValid)
+  {
+    return false;
+  }
+
+  if (timeValid)
+  {
+    switch (compareTimes(lhs.get_time_usec(), rhs.get_time_usec()))
     {
-      return false;
+      case ComparisonResult::False:
+        return false;
+      case ComparisonResult::True:
+        return true;
+      default:
+        break;
     }
-, true)
+  }
 
-COMPARE(>=,
-    return this->get_time_usec() >= rhs.get_time_usec();
-, true)
+  if (frameValid)
+  {
+    return compareFrames(lhs.get_frame(), rhs.get_frame());
+  }
 
-COMPARE(<=,
-    return this->get_time_usec() <= rhs.get_time_usec();
-, true)
+  return fallback;
+}
 
-COMPARE(>,
-    if (this->get_time_usec() > rhs.get_time_usec())
-    {
-      return true;
-    }
-    else if (this->get_time_usec() < rhs.get_time_usec())
-    {
-      return false;
-    }
-, false)
+// ----------------------------------------------------------------------------
+#define COMPARE_TIMES(LOGIC) \
+  [](kv::time_usec_t lhs, kv::time_usec_t rhs){ \
+    do { LOGIC } while(0); \
+    return ComparisonResult::Indeterminate; \
+  }
+#define COMPARE_FRAMES(OP) \
+  [](kv::frame_id_t lhs, kv::frame_id_t rhs){ return lhs OP rhs; }
 
-COMPARE(<,
-    if (this->get_time_usec() < rhs.get_time_usec())
-    {
-      return true;
-    }
-    else if (this->get_time_usec() > rhs.get_time_usec())
-    {
-      return false;
-    }
-, false)
+// ----------------------------------------------------------------------------
+bool TimeStamp::operator==(TimeStamp const& rhs) const
+{
+  return compare(
+    *this, rhs,
+    COMPARE_TIMES(
+      if (lhs != rhs)
+      {
+        return ComparisonResult::False;
+      }
+    ),
+    COMPARE_FRAMES(==),
+    true);
+}
+
+// ----------------------------------------------------------------------------
+bool TimeStamp::operator>=(TimeStamp const& rhs) const
+{
+  return compare(
+    *this, rhs,
+    COMPARE_TIMES(
+      return (lhs >= rhs) ? ComparisonResult::True : ComparisonResult::False;
+    ),
+    COMPARE_FRAMES(>=),
+    true);
+}
+
+// ----------------------------------------------------------------------------
+bool TimeStamp::operator>(TimeStamp const& rhs) const
+{
+  return compare(
+    *this, rhs,
+    COMPARE_TIMES(
+      if (lhs > rhs)
+      {
+        return ComparisonResult::True;
+      }
+      if (lhs < rhs)
+      {
+        return ComparisonResult::False;
+      }
+    ),
+    COMPARE_FRAMES(>),
+    false);
+}
+
+// ----------------------------------------------------------------------------
+bool TimeStamp::operator<=(TimeStamp const& rhs) const
+{
+  return rhs.operator>=(*this);
```

Comment:
Is there a reason why you can't say:

```c++
return rhs > *this;
```
?

Reply:
Probably not...